File: src/main/java/jibril/commands/funny/ShadowDice.java
package jibril.commands.funny;

import java.util.Random;

public class ShadowDice {
    private final Random random = new Random(), shadow = new Random(random.nextLong());

    public int roll(int sides) {
        return Math.min(random.nextInt(sides) + (sides / 5 == 0 ? 0 : shadow.nextInt(sides / 5)), sides - 1) + 1;
    }
}

File: src/main/kotlin/pw/aru/commands/funny/dice/ShadowDice.kt
@file:Suppress("NOTHING_TO_INLINE")

package pw.aru.commands.funny.dice

import pw.aru.utils.AruTaskExecutor.schedule
import java.lang.System.nanoTime
import java.util.*
import java.util.concurrent.TimeUnit.MILLISECONDS

class ShadowDice {
    val r = Random()

    init {
        schedule(65536, MILLISECONDS) { r.setSeed(3447679086515839964L xor nanoTime()) }
    }

    fun roll(sides: Int) = r.run {
        ((nextDouble() * 0.9 + nextDouble() * nextDouble() * nextDouble() * 0.9) * sides)
            .toInt().coerceIn(0, sides - 1).plus(1)
    }
}
